code/krepel/math/vector3.d

```diff
@@ -0,0 +1,569 @@
+    assert(Vec == Vector3.RightVector);
+    // Function
+    assert(Cross(Vector3.UpVector, Vector3.ForwardVector) == Vector3.RightVector);
+    Vector3 Vec1 = Vector3.UpVector;
+    Vector3 Vec2 = Vector3.ForwardVector;
+    // UFCS
+    Vector3 Vec3 = Vec1.Cross(Vec2);
+    assert(Vec3 == Vector3.RightVector);
+  }
+
+  /// Normalization
+  unittest
+  {
+    Vector3 Vec = Vector3(1,1,1);
+    Vec.Normalize();
+    float Expected = 1.0f/Sqrt(3);
+    assert(Vec == Vector3(Expected, Expected, Expected));
+
+    Vec = Vector3(1,1,1);
+    auto Normalized = Vec.NormalizedCopy();
+    auto NormalizedUFCS = NormalizedCopy(Vec);
+    assert(Vec == Vector3(1,1,1));
+    assert(Normalized == Vector3(Expected, Expected, Expected));
+    assert(NormalizedUFCS == Vector3(Expected, Expected, Expected));
+  }
+
+  /// Project Onto Normal
+  unittest
+  {
+    Vector3 Normal = Vector3.UpVector;
+    Vector3 ToProject = Vector3(1,1,0.5f);
+
+    Vector3 Projected = ToProject.ProjectOntoNormal(Normal);
+
+    assert(Projected == Vector3(0,0,0.5f));
+  }
+
+  /// Project Onto PLane
+  unittest
+  {
+    Vector3 Normal = Vector3.UpVector;
+    Vector3 ToProject = Vector3(1,1,0.5f);
+
+    Vector3 Projected = ToProject.ProjectOntoPlane(Normal);
+
+    assert(Projected == Vector3(1,1,0));
+  }
+
+  /// Reflect Vector
+  unittest
+  {
+    Vector3 Normal = Vector3(1,0,0);
+    Vector3 Reflection = Vector3(-1,0,-1);
+
+    Vector3 Reflected = Reflection.ReflectVector(Normal);
+
+    assert(Reflected == Vector3(1,0,-1));
+  }
+
+  /// NearlyEquals
+  unittest
+  {
+    Vector3 A = Vector3(1,1,0);
+    Vector3 B = Vector3(1+1e-5f,1,-1e-6f);
+    Vector3 C = Vector3(1,1,10);
+    assert(NearlyEquals(A,B));
+    assert(!NearlyEquals(A,C));
+  }
+
+  /// Clamp Vector3
+  unittest
+  {
+    Vector3 Vec = Vector3(100,0,0);
+    Vector3 Clamped = Vec.ClampSize(1);
+    assert(Vec == Vector3(100,0,0));
+    assert(Clamped == Vector3(1,0,0));
+  }
+
+  /// Clamp2D Vector3
+  unittest
+  {
+    Vector3 Vec = Vector3(100,0,1000);
+    Vector3 Clamped = Vec.ClampSize2D(1);
+    assert(Vec == Vector3(100,0,1000));
+    assert(Clamped == Vector3(1,0,1000));
+  }
+
+  /// 2D Length
+  unittest
+  {
+    Vector3 Vec = Vector3(0,1,1032094);
+    assert(Vec.Length2D() == 1);
+    assert(Vec.LengthSquared2D() == 1);
+  }
+
+  // Vector3
+  unittest
+  {
+    assert(Vector3(Vector2(1, 2), 3).X == 1);
+    assert(Vector3(Vector2(1, 2), 3).Y == 2);
+    assert(Vector3(Vector2(1, 2), 3).Z == 3);
+
+    assert(Vector3(1, Vector2(2, 3)).X == 1);
+    assert(Vector3(1, Vector2(2, 3)).Y == 2);
+    assert(Vector3(1, Vector2(2, 3)).Z == 3);
+  }
+
+  // Vector3 Swizzle
+  unittest
+  {
+    assert(Vector3(1, 2, 3).ZX   == Vector2(3, 1));
+    assert(Vector3(1, 2, 3).XZX  == Vector3(1, 3, 1));
+    assert(Vector3(1, 2, 3).XYXY == Vector4(1, 2, 1, 2));
+    assert(Vector3(1, 2, 3)._ZYX == Vector3(3, 2, 1));
+    assert(Vector3(1, 2, 3)._0YX == Vector3(0, 2, 1));
```

Comment:
Add a test for `0` inbetween without `_`? Like `Vector3(1, 2, 3).X0Z == Vector3(1, 0, 3)`.